### ISSUE TITLE
Installing Google Play Services leads to dex files always being recompiled

### ISSUE THREAD

**Issue by ElteHupkes** [xamarin/xamarin-android#1256]
I have to admit I'm not certain where to report this issue, since it's Android related I figured I'd try here.

### Steps to Reproduce

1. In VS for Mac (haven't tried on Windows so far), create a new, blank Xamarin.Android app. 
2. Build the project. Build it again to note the build time when the project is already built.
3. Install the `Xamarin.GooglePlayServices.Basement` NuGet package, or just any GooglePlayServices package that uses it.
4. Rebuild the project. Build it again, once more noting the build time.

### Expected Behavior
In both (2) and (4) the second builds should be pretty much instantaneous, since there are no changes and build tasks are all being skipped because of it.

### Actual Behavior
Indeed (2) has the expected behavior. In (4), however, any build executes the `CompileToDalvik` task, which takes several seconds on the empty app (and much longer on my real app). What's worse is that the dex files generated because of it cause a redeployment of the app every time it is run, significantly increasing deploy / run times. 

From the diagnostic build output:
```
Target "_CompileToDalvikWithDx" in file "/Library/Frameworks/Mono.framework/External/xbuild/Xamarin/Android/Xamarin.Android.Common.targets":
  Building target "_CompileToDalvikWithDx" completely.
  Input file "obj/Debug/XbdMerge/proguard/dc663e480.txt" is newer than output file "obj/Debug/_dex_stamp".
```

That file, in turn, appears to be generated by the `_XamarinAndroidBuildAarProguardConfigs` target. I haven't managed to figure out what's going on here exactly, but that file seems to disappear at the start of every build.

### Version Information
=== Visual Studio Community 2017 for Mac ===

Version 7.3.3 (build 7)
Installation UUID: ccf6f6c1-ad58-4aa7-849b-9a9ced210760
Runtime:
	Mono 5.4.1.7 (2017-06/e66d9abbb27) (64-bit)
	GTK+ 2.24.23 (Raleigh theme)

	Package version: 504010007

=== NuGet ===

Version: 4.3.1.4445

=== .NET Core ===

Runtime: /usr/local/share/dotnet/dotnet
Runtime Version: 2.0.0
SDK: /usr/local/share/dotnet/sdk/2.0.2/Sdks
SDK Versions:
	2.0.2
	2.0.0
MSBuild SDKs: /Library/Frameworks/Mono.framework/Versions/5.4.1/lib/mono/msbuild/15.0/bin/Sdks

=== Xamarin.Profiler ===

Version: 1.6.0
Location: /Applications/Xamarin Profiler.app/Contents/MacOS/Xamarin Profiler

=== Apple Developer Tools ===

Xcode 9.2 (13772)
Build 9C40b

=== Xamarin.iOS ===

Version: 11.6.1.4 (Visual Studio Community)
Hash: db807ec9
Branch: xcode9.2
Build date: 2018-01-10 16:45:48-0500

=== Xamarin.Android ===

Version: 8.1.5.0 (Visual Studio Community)
Android SDK: /Users/elte/Library/Android/sdk
	Supported Android versions:
		4.4 (API level 19)
		6.0 (API level 23)
		7.1 (API level 25)
		8.0 (API level 26)

SDK Tools Version: 26.1.1
SDK Platform Tools Version: 27.0.1
SDK Build Tools Version: 27.0.3

Java SDK: /usr
java version "1.8.0_131"
Java(TM) SE Runtime Environment (build 1.8.0_131-b11)
Java HotSpot(TM) 64-Bit Server VM (build 25.131-b11, mixed mode)

Android Designer EPL code available here:
https://github.com/xamarin/AndroidDesigner.EPL

=== Xamarin.Mac ===

Version: 4.0.0.216 (Visual Studio Community)

=== Xamarin Inspector ===

Version: 1.4.0
Hash: b3f92f9
Branch: master
Build date: Fri, 19 Jan 2018 22:00:34 GMT
Client compatibility: 1

=== Build Information ===

Release ID: 703030007
Git revision: 125911fb4accc4309b2cee5c81c970c7cff9b0e0
Build date: 2018-01-22 17:46:46-05
Xamarin addins: 463e21a6d9d4f6b57f923df376fff093a1dd9404
Build lane: monodevelop-lion-d15-5

=== Operating System ===

Mac OS X 10.13.2
Darwin 17.3.0 Darwin Kernel Version 17.3.0
    Thu Nov  9 18:09:22 PST 2017
    root:xnu-4570.31.3~1/RELEASE_X86_64 x86_64

=== Enabled user installed addins ===

Internet of Things (IoT) development (Preview) 7.1

### Log File
Android log not applicable, I'll paste the build log here if needed (it's really long with diagnostics).

**Comment by atsushieno** [comment]
I don't think expecting unchanged build is right. Installing NuGet packages means that your .csproj or .fsproj is modified, resulting in any possible code (DLL) changes and Java bytecode changes. If we don't update dex, that will result in inconsistent app packaging and unexpected run-time breakage.

If it is triggered by repetitive build operations *after* installing NuGet packages (i.e. nuget install -> build -> build), then that's a valid expectation; such builds should not result in re-compilation of dexes.

**Comment by ElteHupkes** [comment]
I'm not talking about the rebuilding after installing NuGet packages - that's perfectly logical. It rebuilds, however, if I do absolutely nothing at all. Press build, wait until it finishes, press build again. Before installing the packages this works as expected. After installing the packages, repeated builds recompile every time without a single change being made to anything at all.

Upon further debugging I've managed to trace it down to this commit in `Xamarin.Build.Download`: https://github.com/xamarin/XamarinComponents/commit/ffdd6cf4b1dbf572ea5b1068fb24b80e876208ef#diff-3334d0a45edc363a49d7bba924be58b3
which also explains why I've only started to experience this problem recently, because we only recently updated to `Xamarin.Build.Download` 0.4.7 that contains this commit. The `AdditionalFileWrites` added in this commit end up in the `[Project]/obj/Debug/[Project].csproj.FileListAbsolute.txt.`, and those files are removed before every build. Since this contains the `.proguard.stamp` files the `XamarinBuildAarProguardConfigs` task recreates its proguard files, which in turn results in a recompilation of the dex files. Indeed if I remove those lines from that `FileListAbsolute` file recompilation does not happen (I'm using that as a workaround for now).

I don't really understand how these `AdditionalFileWrites` end up in that file and how that relates to the build process, but given that this is caused by a change in `Xamarin.Build.Download`, should I file the issue there instead?

**Comment by ElteHupkes** [comment]
After a bit more investigation I think I now know in more detail what's going on. There are 3 relevant build targets involved: 

- `XamarinBuildAndroidAarProguardConfigs`,
- `IncrementalClean`
- `CompileToDalvik`

They run in that order. `XamarinBuildAndroidAarProguardConfigs` creates some files in `obj/Debug/XbdMerge/` and lists these as target `FileWrites`, and as such they end up in the `[csproj].FileListAbsolute.txt`. `IncrementalClean` looks at `FileWrites` that were outputs of the last build but not the current one, and deletes these files (to get rid of, for instance, outputs of renamed libraries). When `XamarinBuildAndroidAarProguardConfigs` notes the presence of a `[hash].proguard.config` for any particular config it's trying to create, it skips that task and, crucially, _does not add the corresponding files to the `FileWrites`_. Because of this, when `IncrementalClean` runs it notes that these proguard files were a part of the previous build but not of this build, and deletes them. `CompileToDalvik` runs because its input files (the proguard files) are missing. On the next build, the files are regenerated.

In conclusion, before a build process the files are either not there yet and will be generated (causing `CompileToDalvik` to run because they're newer), or they're already there but deleted by `IncrementalClean` (causing `CompileToDalvik` to run because they're missing). The `CompileToDalvik` task runs every time, and on alternating builds the files are either present or missing after the build.

I think this is an issue with `Xamarin.Build.Download` and will proceed to report it there.

**Comment by dellis1972** [comment]
@ElteHupkes we have been having allot of problems with IncrementalClean recently. It does sound like the problem is that the FileWrites ItemGroup and consequently the FileListAbsolute.txt file is missing some entries which should be kept around. @Redth this looks like something in your area. We should probably update those tasks to always include the TaskItems if they exist on disk , regardless of if they were created in this build cycle.

**Comment by ElteHupkes** [comment]
@dellis1972 That was my hunch as well, but then again I can't claim to have any experience whatsoever regarding how this build process operates. Should I still "move" this issue to XamarinComponents or do you prefer to track it here?

**Comment by dellis1972** [comment]
Lets see what @Redth thinks :)

**Comment by Redth** [comment]
Ok so I think what's happening here is when we process the proguard config files from .aar's downloaded through Xamarin.Build.Download, we keep a .stamp file around to help with caching things.

https://github.com/xamarin/XamarinComponents/blob/master/Util/Xamarin.Build.Download/source/Xamarin.Build.Download/XamarinBuildAndroidAarProguardConfigs.cs#L64-L66

However, if we detect the stamp file exists, we aren't emitting the filenames of the actual proguard config files we previously cached (or the stamp file itself for that matter).  This was fine for our own task's Inputs/Outputs, but it didn't take into consideration that the proguard config files were being added to another item group which apparently does care about these items being in FileWrites.

Seems like the fix will be to add any proguard config files we previously cached to FileWrites on every task invocation before returning from the task.

**Comment by dellis1972** [comment]
@Redth it looks like you fixed this upstream, can we close this issue?

**Comment by Redth** [comment]
Yes this is fixed in 60.1142.1
